subsys/sd/sd_ops.c

```diff
@@ -63,21 +63,28 @@ int sdmmc_read_status(struct sd_card *card)
 int sdmmc_wait_ready(struct sd_card *card)
 {
 	int ret, timeout = CONFIG_SD_DATA_TIMEOUT * 1000;
-	bool busy = true;
 
 	do {
-		busy = sdhc_card_busy(card->sdhc);
-		if (!busy) {
+		if (!sdhc_card_busy(card->sdhc)) {
 			/* Check card status */
 			ret = sd_retry(sdmmc_read_status, card, CONFIG_SD_RETRY_COUNT);
-			busy = (ret != 0);
-		} else {
-			/* Delay 125us before polling again */
-			k_busy_wait(125);
-			timeout -= 125;
+			if (ret == 0) {
+				return 0;
+			}
+			if (ret == -ETIMEDOUT) {
+				/* If this check timed out, then the total
+				 * time elapsed in microseconds is
+				 * SD_CMD_TIMEOUT * SD_RETRY_COUNT * 1000
+				 */
+				timeout -= (CONFIG_SD_CMD_TIMEOUT *
+					    CONFIG_SD_RETRY_COUNT) * 1000;
+			}
 		}
-	} while (busy && (timeout > 0));
-	return busy;
+		/* Delay 125us before polling again */
+		k_busy_wait(125);
+		timeout -= 125;
+	} while (timeout > 0);
```

Comment:
`CONFIG_SD_CMD_TIMEOUT` being greater than ~22 seconds with `CONFIG_SD_RETRY_COUNT` something like 100 could result in integer overflow in this code, I'm not sure if this is a reasonably expected scenario though, it seems a little ridiculous, but I'm just pointing this out.

Reply:
Just in case if you think it's an issue you could use [u32_add_overflow](https://docs.zephyrproject.org/latest/doxygen/html/group__math__extras.html#ga3f36aa272f5595d78c6e8219b2c4dbfb) from math_extras.h

Reply:
```
$ west build -p always -b lpcxpresso55s69_cpu0 tests/drivers/disk/disk_access/ -DCONFIG_SD_CMD_TIMEOUT=22000 -DCONFIG_SD_RETRY_COUNT=10

[99/155] Building C object zephyr/subsys/sd/CMakeFiles/subsys__sd.dir/sd_ops.c.obj
/home/danieldegrasse/zephyr/zephyrproject/zephyr/subsys/sd/sd_ops.c: In function 'sdmmc_wait_ready':
/home/danieldegrasse/zephyr/zephyrproject/zephyr/subsys/sd/sd_ops.c:80:68: warning: integer overflow in expression of type 'int' results in '-2094967296' [-Woverflow]
   80 |                                             CONFIG_SD_RETRY_COUNT) * 1000;
```
Given that the compiler warns about it, I think we are ok to leave this as is.